# bioinformatics in Ukrainian - Біоінформатика українською

* [Ukrainian MOH - by Inna Dubchak - 4 video](https://www.youtube.com/watch?v=HsBBIJEkf3k): Intro, Alignment, Phylogenies, Blast
* [GenomicsUA](https://www.youtube.com/c/NGOGenomicsUA/videos)
* [Сбірка геномів](https://ukr.legatechnics.com/de-novo-genome-assembly-what-every-biologist-should-know-53879722)

# tutorials I taught

* [**The basics of IGV**](https://docs.google.com/presentation/d/1dUNRB76aZpQgJjPqR_nhzPKztfd9uV5biOTprONc7MU/edit?usp=sharing)\
Session 8 in DNA-seq bootcamp 2019, CCM, Hospital for Sick Children, Toronto, 2019-03-15.
* [**Real time genomics analysis with iobio.io toolkit**](https://docs.google.com/presentation/d/1PdSo2VvHyMDbiX_tMLWkxOYLCpYE4dA0HCjmg56LCmo/edit?usp=sharing])\
Session 8 in DNA-seq bootcamp 2019, CCM, Hospital for Sick Children, Toronto, 2019-03-15.
* [**Small and structural variant prioritization in R with tidyverse**](https://docs.google.com/presentation/d/1SBeRCSlwfsLOLvx5fSDHd1sM2Hu3lwgDSB__TAduuSg/edit?usp=sharing)\
Session 10 in DNA-seq bootcamp 2019, CCM, Hospital for Sick Children, Toronto, 2019-03-22.

## Variant annotation and prioritization series
* [Small and structural variant prioritization in R with tidyverse](https://docs.google.com/presentation/d/1SBeRCSlwfsLOLvx5fSDHd1sM2Hu3lwgDSB__TAduuSg/edit?usp=sharing)
* **Variant calling and annotation** ([gist](https://gist.github.com/naumenko-sa/82df1cb7d9b5f64691bf437f0eb455f0), [slides](https://drive.google.com/open?id=158TJQTiluXCbTEcasNp0WfTylhykwtAk)).\
Genomics in Medicine bioinformatics.ca workshop, module 3, McGill, Montreal, 2018-06-13. \
~4h tutorial: small bcbio WES run, VEP, gemini, gemini queries, HPC, excel reports, variants prioritization, IGV. Bonus: how to become a bioinf pro with 4 tools.
* **Elements of variant analysis.**\
CCM bioinformatics tutorials for biologists, Hospital for Sick Children, Toronto, 2018-03-16.\
IGV, iobio.io toolkit, vcf>vep>gemini, bcbio example run.\
*Obsolete, developed into IGV, iobio, and Montreal tutorials*
* [Visualization of bam, vcf (QC), gene set enrichment analysis of RNA-seq data](https://docs.google.com/presentation/d/1RRnTsTWYn-Yo5OahbFj4F4ViWPuHIvMkzfYrOqxutv0/edit?usp=sharing)\
CCM bioinformatics tutorial for biologists, Hospital for Sick Children, Toronto, 2017-04-28.
IGV, iobio.io toolkit, a bit of RNA-seq and GSEA.
* [**Exploring small variants with GEMINI database in R**](https://docs.google.com/presentation/d/1DQ0dInX2VXSYE137nOzgeFKVOv5Ww5OBIENjBZRQDr0/edit?usp=sharing).
CCM bioinformatics tutorials for biologists, Hospital for Sick Children, Toronto, 2016-11-30, updated in 2019.\
R, R-studio, gemini.db, RSQLite, tidyverse, bcbio.

## BiomaRt series
* [**Genomics in R: genes, regions, variants (biomaRt, VariantAnnotation, TopMed)**](https://docs.google.com/presentation/d/1rTG-FDQ9y8c8e6cO1WdVwq0_BF_R1VRrjTgUD1Vmv54/edit?usp=sharing).\
A Guest lecture for UofT Medical Genomics program, UofT, Toronto, 019-01-30.
* Genomics in R: genes, regions, variants: CCM tutorial, SickKids, Toronto, 2018-11-21.
* Genomics in R: genes, regions, variants: CCM tutorial, SickKids, Toronto, 2017-11-21.

## Out of series
* [*De novo* genome assembly: a practical guide (in Russian)](http://makarich.fbb.msu.ru/snaumenko/ngs_lecture/naumenko.genome_assembly-n.pdf).
A tutorial for Systems Biology course organized by M. Gelfand, Moscow State University, Moscow, 2014 and 2015.

# tutorials, courses, and books I found useful
* [Applied Computational Genomics by Aaron Quinlan](https://github.com/quinlan-lab/applied-computational-genomics)
* [HBC training tutorials](https://github.com/hbctraining)
* [Data Visualization by Claus O. Wilke](https://serialmentor.com/dataviz/)
* [Evolutionary Biology by AS Kondrashov](https://www.youtube.com/watch?v=fKv2UBkfmA0&list=PLfMMXtHYUY1r6ONv8pzOXCIt5BzH7gxPs)
